marxs/optics/multiLayerMirror.py

```diff
@@ -0,0 +1,104 @@
+        p_v_2 = np.zeros(len(polarization))
+        for i in range(0, len(polarization)):
+        	v_1[i] /= np.linalg.norm(v_1[i])
+        	v_2[i] /= np.linalg.norm(v_2[i])
+        	p_v_1[i] = np.dot(polarization[i], v_1[i])   # just the cosine between the vectors
+        	p_v_2[i] = np.dot(polarization[i], v_2[i])
+        
+        # reflect (change direction)
+        directions = photons['dir']
+        pos4d_inv = np.linalg.inv(self.pos4d)
+        directions = directions.T
+        directions = np.dot(pos4d_inv, directions)
+        directions[0,:] *= -1
+        directions = np.dot(self.pos4d, directions)
+        directions = directions.T
+        photons['dir'] = directions
+        
+        # adjust polarization vectors after reflection
+        # find new v_2
+        new_beam_dir = (photons['dir'][:,0:3]).copy()
+        for i in range(0, len(new_beam_dir)):
+        	new_beam_dir[i] /= np.linalg.norm(new_beam_dir[i])
+        new_v_2 = np.cross(new_beam_dir, v_1)
+        for i in range(0, len(polarization)):
+        	photons['polarization'][i,0:3] = p_v_1[i] * v_1[i] + p_v_2[i] * new_v_2[i]
+        
+        # set position to intersection point
+        photons['pos'] = intersection
+        
+        # read in correct reflecting probability file, now in table format
+    	reflectFile = ascii.read(self.fileName)
+        
+        # find reflectivity adjustment due to polarization in testing
+        polarizedFile = ascii.read('./marxs/optics/docs/ALSpolarization2.txt')
+        tested_polarized_fraction = np.interp(photons['energy'], polarizedFile['Photon energy'] / 1000, polarizedFile['Polarization'])
+        
+        # find probability of being reflected due to position
+        local_intersection = h2e((np.dot(pos4d_inv, intersection.T)).T)
+        local_coords_in_file = reflectFile['X(mm)'] / np.linalg.norm(self.geometry['v_y']) - 1
+        peak_wavelength = np.interp(local_intersection[:,1], local_coords_in_file, reflectFile['Peak lambda'])
+        max_refl = np.interp(local_intersection[:,1], local_coords_in_file, reflectFile['Peak']) / tested_polarized_fraction
+        spread_refl = np.interp(local_intersection[:,1], local_coords_in_file, reflectFile['FWHM(nm)'])
+        
+        wavelength = 1.23984282 / photons['energy']   # wavelength is in nm assuming energy is in keV
+        c_squared = (spread_refl ** 2) / (2 * np.log(2))
+        refl_prob = max_refl * np.exp(-((wavelength - peak_wavelength) ** 2) / (2 * c_squared))
+        
+        # find probability of being reflected due to polarization
+        # v_1 is parallel to plane, good reflection direction
+        for i in range(0, len(polarization)):
+        	refl_prob[i] *= np.dot(polarization[i], v_1[i])**2
+        
+        # multiply probability by probability of reflection
+        for i in range(0, len(photons['probability'])):
+        	if not (photons['probability'][i] == 0):
+        		photons['probability'][i] *= refl_prob[i] / 100
```

Comment:
I think these three lines can be simplified to:
`photons['probability'][i] *= refl_prob[i] / 100`. Multiplying 0 with anything will give 0 again...
In C or Fortran would would be right to avoid the * and / for those operations where it does not matter. In Python, I suspected that saving the for loop is more important, but I don't really know, so I run some tests:
```
In [63]: prob = np.random.rand(i)

In [64]: refl = np.random.rand(i)
In [78]: def t(prob, refl, p):
    for i in range(len(prob)):
        if prob[i] != 0:
   ....:             p[i] = prob[i] * refl[i] /100
   ....:             


In [80]: p = np.zeros_like(prob)

In [81]: %timeit t(prob, refl,p)
10 loops, best of 3: 69.3 ms per loop

In [83]: %timeit p = prob * refl /100
1000 loops, best of 3: 228 µs per loop
# Your version will be better if prob has more 0 in it:
In [84]: prob[::2] = 0

In [85]: %timeit t(prob, refl,p)
10 loops, best of 3: 51.8 ms per loop
```
In practice, your version will be even slower, because it needs to go through the lookup of `photons['probability']` many more times than my version.

Don't worry though, there is no way you could have predicted this from your C experience - this is really specific to interpreted languages like python, where often much of the speed gain is in "use the package that implements it in C" instead of C where you do the optimizing yourself.